<img src="https://raw.githubusercontent.com/andre-marcos-perez/ebac-course-utils/main/media/logo/newebac_logo_black_half.png" alt="ebac-logo">

---

# **Módulo** | Análise de Dados: COVID-19 Dashboard
Caderno de **Exercícios**<br> 
Professor [André Perez](https://www.linkedin.com/in/andremarcosperez/)

---

# **Tópicos**

<ol type="1">
  <li>Introdução;</li>
  <li>Análise Exploratória de Dados;</li>
  <li>Visualização Interativa de Dados;</li>
  <li>Storytelling.</li>
</ol>

---

# **COVID Dashboard**

## 1\. Introdução

### **1.1. TLDR** 

 - **Dashboard**: 
  - Google Data Studio ([link](https://lookerstudio.google.com/reporting/13a73b4c-ee10-4774-8746-a22608a8839c)).
 - **Processamento**:
  - Kaggle Notebook ([link](https://www.kaggle.com/caiomarquesribeiro/covid-dashboard)). 
 - **Fontes**: 
  - Casos pela universidade John Hopkins ([link](https://github.com/CSSEGISandData/COVID-19/tree/master/csse_covid_19_data/csse_covid_19_daily_reports));
  - Vacinação pela universidade de Oxford ([link](https://covid.ourworldindata.org/data/owid-covid-data.csv)).

### **1.2. Pandemia Coronavírus

> A COVID-19 é uma infecção respiratória aguda causada pelo coronavírus SARS-CoV-2, potencialmente grave, de elevada transmissibilidade e de distribuição global. Fonte: Governo brasileiro ([link](https://www.gov.br/saude/pt-br/coronavirus/o-que-e-o-coronavirus)).

Para melhor compreensão do processo de vacinação, este projeto busca construir um dashboard de dados para exploração e visualização interativa de dados sobre o avanço de casos e da vacinação do Brasil entre o periodo de jan/2021 a dez/2022. O processamento de dados está neste [link](https://www.kaggle.com/caiomarquesribeiro/covid-dashboard) e o dashboard, neste [link](https://lookerstudio.google.com/reporting/13a73b4c-ee10-4774-8746-a22608a8839c).

### **1.3. Dados** 

Os dados sobre **casos da COVID-19** são compilados pelo centro de ciência de sistemas e engenharia da universidade americana **John Hopkins** ([link](https://www.jhu.edu)). Os dados são atualizados diariamente deste janeiro de 2020 com uma granularidade temporal de dias e geográfica de regiões de países (estados, condados, etc.). O website do projeto pode ser acessado neste [link](https://systems.jhu.edu/research/public-health/ncov/) enquanto os dados, neste [link](https://github.com/CSSEGISandData/COVID-19/tree/master/csse_covid_19_data/csse_covid_19_daily_reports). Abaixo estão descritos os dados derivados do seu processamento.

 - **date**: data de referência;
 - **state**: estado;
 - **country**: país; 
 - **population**: população estimada;
 - **confirmed**: número acumulado de infectados;
 - **confirmed_1d**: número diário de infectados;
 - **confirmed_moving_avg_7d**: média móvel de 7 dias do número diário de infectados;
 - **confirmed_moving_avg_7d_rate_14d**: média móvel de 7 dias dividido pela média móvel de 7 dias de 14 dias atrás;
 - **deaths**: número acumulado de mortos;
 - **deaths_1d**: número diário de mortos;
 - **deaths_moving_avg_7d**: média móvel de 7 dias do número diário de mortos;
 - **deaths_moving_avg_7d**: média móvel de 7 dias dividido pela média móvel de 7 dias de 14 dias atrás;
 - **month**: mês de referência;
 - **year**: ano de referência.

Os dados sobre **vacinação da COVID-19** são compilados pelo projeto Nosso Mundo em Dados (*Our World in Data* ou OWID) da universidade britânica de **Oxford** ([link](https://www.ox.ac.uk)). Os dados são **atualizados diariamente** deste janeiro de 2020 com uma **granularidade temporal de dias e geográfica de países**. O website do projeto pode ser acessado neste [link](https://ourworldindata.org) enquanto os dados, neste [link](https://covid.ourworldindata.org/data/owid-covid-data.csv). Abaixo estão descritos os dados derivados do seu processamento.

 - **date**: data de referência;
 - **country**: país;
 - **population**: população estimada;
 - **total**: número acumulado de doses administradas;
 - **one_shot**: número acumulado de pessoas com uma dose;
 - **one_shot_perc**: número acumulado relativo de pessoas com uma dose;
 - **two_shots**: número acumulado de pessoas com duas doses;
 - **two_shot_perc**: número acumulado relativo de pessoas com duas doses;
 - **three_shots**: número acumulado de pessoas com três doses;
 - **three_shot_perc**: número acumulado relativo de pessoas com três doses;
 - **month**: mês de referência;
 - **year**: ano de referência.

## 2\. Análise Exploratória de Dados

Nesta sessão vamos utilizar os seguintes pacotes Python para processar os dados bruto em um formato adequado para um painel para exploração interativa de dados.

In [2]:
import math
from typing import Iterator
from datetime import datetime, timedelta

import numpy as np
import pandas as pd

### **2.1. Casos** 

Vamos processar os dados de **casos** da universidade John Hopkins.



#### **2.1.1. Extração**

O dado está compilado em um arquivo por dia, exemplo para 2021/12/01.

In [3]:
cases = pd.read_csv('https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_daily_reports/01-12-2021.csv', sep=',')

In [4]:
cases.head()

FIPS Admin2 Province_State Country_Region          Last_Update       Lat   
0   NaN    NaN            NaN    Afghanistan  2021-01-13 05:22:15  33.93911  \
1   NaN    NaN            NaN        Albania  2021-01-13 05:22:15  41.15330   
2   NaN    NaN            NaN        Algeria  2021-01-13 05:22:15  28.03390   
3   NaN    NaN            NaN        Andorra  2021-01-13 05:22:15  42.50630   
4   NaN    NaN            NaN         Angola  2021-01-13 05:22:15 -11.20270   

       Long_  Confirmed  Deaths  Recovered  Active Combined_Key   
0  67.709953      53584    2301      44608    6675  Afghanistan  \
1  20.168300      64627    1252      38421   24954      Albania   
2   1.659600     102641    2816      69608   30217      Algeria   
3   1.521800       8682      86       7930     666      Andorra   
4  17.873900      18343     422      15512    2409       Angola   

   Incident_Rate  Case_Fatality_Ratio  
0     137.647787             4.294192  
1    2245.708527             1.937271  
2     234.067409             2.743543  
3   11236.653077             0.990555  
4      55.811022             2.300605

Portanto, precisaremos iterar dentro de um intervalo de tempo definido para extraí-lo.

In [6]:
# A função date_range tem como meta a iteração de cada data entre 2 periodos, start_date e end_date, ambos informados manualmente.
def date_range(start_date: datetime, end_date: datetime) -> Iterator[datetime]:
  date_range_days: int = (end_date - start_date).days
  for lag in range(date_range_days):
    yield start_date + timedelta(lag)

In [7]:
start_date = datetime(2021,  1,  1)
end_date   = datetime(2022, 12, 31)

De maneira iterativa, vamos selecionar as colunas de interesse e as linhas referentes ao Brasil.

In [8]:
cases = None
cases_is_empty = True

# Para cada data iterada entre a data inicial e final, será importada pela variavél data_source_url o .csv da data respectiva.
for date in date_range(start_date=start_date, end_date=end_date):

  date_str = date.strftime('%m-%d-%Y')
  data_source_url = f'https://raw.githubusercontent.com/CSSEGISandData/COVID-19/master/csse_covid_19_data/csse_covid_19_daily_reports/{date_str}.csv'

  case = pd.read_csv(data_source_url, sep=',')

  # Haverá a separação dos dados que serão utilizados para essa leitura, dropando os dados sem relevância no momento.
  case = case.drop(['FIPS', 'Admin2', 'Last_Update', 'Lat', 'Long_', 'Recovered', 'Active', 'Combined_Key'], axis=1)
  case = case.query('Country_Region == "Brazil"').reset_index(drop=True)

  # É importante manter a conversão da data padronizada de acordo com o datetime, para importação futura no Data Studio.
  case['Date'] = pd.to_datetime(date.strftime('%Y-%m-%d'))

  # If/Else para diferenciar a ação entre o primeiro elemento iterado e os outros, pois varia para a criação do arquivo.
  if cases_is_empty:
    cases = case
    cases_is_empty = False
  else:
    cases = pd.concat([cases, case], ignore_index=True)

In [9]:
cases.query('Province_State == "Sao Paulo"').head()

Province_State Country_Region  Confirmed  Deaths  Incident_Rate   
24       Sao Paulo         Brazil    1466191   46775    3192.990778  \
51       Sao Paulo         Brazil    1467953   46808    3196.827966   
78       Sao Paulo         Brazil    1471422   46845    3204.382565   
105      Sao Paulo         Brazil    1473670   46888    3209.278136   
132      Sao Paulo         Brazil    1486551   47222    3237.329676   

     Case_Fatality_Ratio       Date  
24              3.190239 2021-01-01  
51              3.188658 2021-01-02  
78              3.183655 2021-01-03  
105             3.181716 2021-01-04  
132             3.176615 2021-01-05

#### **2.1.2. Wrangling**

Vamos manipular os dados para o dashboard. O foco é em garantir uma boa granularidade e qualidade da base de dados.

In [10]:
cases.head()

Province_State Country_Region  Confirmed  Deaths  Incident_Rate   
0           Acre         Brazil      41689     796    4726.992352  \
1        Alagoas         Brazil     105091    2496    3148.928928   
2          Amapa         Brazil      68361     926    8083.066602   
3       Amazonas         Brazil     201574    5295    4863.536793   
4          Bahia         Brazil     494684    9159    3326.039611   

   Case_Fatality_Ratio       Date  
0             1.909377 2021-01-01  
1             2.375084 2021-01-01  
2             1.354574 2021-01-01  
3             2.626827 2021-01-01  
4             1.851485 2021-01-01

In [11]:
cases.shape

(19683, 7)

In [12]:
# Dados bastante consolidados no geral, sem valores nulos.
cases.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19683 entries, 0 to 19682
Data columns (total 7 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Province_State       19683 non-null  object        
 1   Country_Region       19683 non-null  object        
 2   Confirmed            19683 non-null  int64         
 3   Deaths               19683 non-null  int64         
 4   Incident_Rate        19683 non-null  float64       
 5   Case_Fatality_Ratio  19683 non-null  float64       
 6   Date                 19683 non-null  datetime64[ns]
dtypes: datetime64[ns](1), float64(2), int64(2), object(2)
memory usage: 1.1+ MB


Começamos com o nome das colunas.

In [13]:
cases = cases.rename(
  columns={
    'Province_State': 'state',
    'Country_Region': 'country',
    
  }
)

for col in cases.columns:
  cases = cases.rename(columns={col: col.lower()})

Ajustamos o nome dos estados.

In [14]:
# Essa mudança na nomeclatura se deve para a adaptação no Goggle Data Studio, que utiliza o Google Maps como base.
states_map = {
    'Amapa': 'Amapá',
    'Ceara': 'Ceará',
    'Espirito Santo': 'Espírito Santo',
    'Goias': 'Goiás',
    'Para': 'Pará',
    'Paraiba': 'Paraíba',
    'Parana': 'Paraná',
    'Piaui': 'Piauí',
    'Rondonia': 'Rondônia',
    'Sao Paulo': 'São Paulo'
}

# Ele vai verificar estado por estado no dataframe, para substituir caso a chave esteja presente na variável states_map.
cases['state'] = cases['state'].apply(lambda state: states_map.get(state) if state in states_map.keys() else state)

Vamos então computar novas colunas para enriquecer a base de dados.

 - Chaves temporais:

In [15]:
cases['month'] = cases['date'].apply(lambda date: date.strftime('%Y-%m'))
cases['year']  = cases['date'].apply(lambda date: date.strftime('%Y'))

 - População estimada do estado:

In [16]:
cases['population'] = round(100000 * (cases['confirmed'] / cases['incident_rate']))
cases = cases.drop('incident_rate', axis=1)

 - Número, média móvel (7 dias) e estabilidade (14 dias) de casos e mortes por estado:

| 1 | 2 | 3 | 4 | 5 | 6 | <font color='red'>7</font> | <font color='green'>8</font> | 9 | 10 | 11 | 12 | 13 | <font color='blue'>14<font color='red'> | 15 | 16 | 17 | 18 | 19 | 20 | 21 |
| - | - | - | - | - | - | - | - | - | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- |
| <font color='red'>D-6</font> | <font color='red'>D-5</font> | <font color='red'>D-4</font> | <font color='red'>D-3</font> | <font color='red'>D-2</font> | <font color='red'>D-1</font> | <font color='red'>D0</font> | | | | | | | | | | | | | | |
| D-7 | <font color='green'>D-6</font> | <font color='green'>D-5</font> | <font color='green'>D-4</font> | <font color='green'>D-3</font> | <font color='green'>D-2</font> | <font color='green'>D-1</font> | <font color='green'>D0</font> | | | | | | | | | | | | | |
| D-13 | D-12 | D-11 | D-10 | D-9 | D-8 | D-7 | <font color='blue'>D-6</font> | <font color='blue'>D-5</font> | <font color='blue'>D-4</font> | <font color='blue'>D-3</font> | <font color='blue'>D-2</font> | <font color='blue'>D-1</font> | <font color='blue'>D0</font> | | | | | | | |

In [17]:
cases_ = None
cases_is_empty = True

# get_trend serve para perceber se a média móvel aumentou, diminuiu ou se manteve estável
def get_trend(rate: float) -> str:

  if np.isnan(rate):
    return np.NaN

  if rate < 0.75:
    status = 'downward'
  elif rate > 1.15:
    status = 'upward'
  else:
    status = 'stable'

  return status
  
# O laço for vai selecionar apenas o estado na coluna cases sem repetir, um de cada vez.
for state in cases['state'].drop_duplicates():

  # Separa-se o estado de interesse e ordena de acordo com a data.
  cases_per_state = cases.query(f'state == "{state}"').reset_index(drop=True)
  cases_per_state = cases_per_state.sort_values(by=['date'])

  # O .diff define a diferença de confirmados em um periodo de um dia
  cases_per_state['confirmed_1d'] = cases_per_state['confirmed'].diff(periods=1)
  # O .rolling pega 1 janela de 7 itens (no caso, em dias) em confirmed_1d e .mean() a média
  cases_per_state['confirmed_moving_avg_7d'] = np.ceil(cases_per_state['confirmed_1d'].rolling(window=7).mean())
  # Aqui acontece a comparação entre a média móvel do dia e a média móvel de 14 periodos atrás. utilizando o .shift
  cases_per_state['confirmed_moving_avg_7d_rate_14d'] = cases_per_state['confirmed_moving_avg_7d']/cases_per_state['confirmed_moving_avg_7d'].shift(periods=14)
  # Por fim, será aplicado o get_trend
  cases_per_state['confirmed_trend'] = cases_per_state['confirmed_moving_avg_7d_rate_14d'].apply(get_trend)

  # O mesmo passo a passo dos casos aplicados para as mortes
  cases_per_state['deaths_1d'] = cases_per_state['deaths'].diff(periods=1)
  cases_per_state['deaths_moving_avg_7d'] = np.ceil(cases_per_state['deaths_1d'].rolling(window=7).mean())
  cases_per_state['deaths_moving_avg_7d_rate_14d'] = cases_per_state['deaths_moving_avg_7d']/cases_per_state['deaths_moving_avg_7d'].shift(periods=14)
  cases_per_state['deaths_trend'] = cases_per_state['deaths_moving_avg_7d_rate_14d'].apply(get_trend)

  if cases_is_empty:
    cases_ = cases_per_state
    cases_is_empty = False
  else:
    cases_ = pd.concat([cases_, cases_per_state], ignore_index=True)

cases = cases_
cases_ = None

Garantir o tipo do dado é fundamental para consistência da base de dados. Vamos fazer o *type casting* das colunas.

In [18]:
cases['population'] = cases['population'].astype('Int64')
cases['confirmed_1d'] = cases['confirmed_1d'].astype('Int64')
cases['confirmed_moving_avg_7d'] = cases['confirmed_moving_avg_7d'].astype('Int64')
cases['deaths_1d'] = cases['deaths_1d'].astype('Int64')
cases['deaths_moving_avg_7d'] = cases['deaths_moving_avg_7d'].astype('Int64')
cases['case_fatality_ratio'] = round(cases['case_fatality_ratio'], 2)

In [19]:
cases.head()

state country  confirmed  deaths  case_fatality_ratio       date    month   
0  Acre  Brazil      41689     796                 1.91 2021-01-01  2021-01  \
1  Acre  Brazil      41941     798                 1.90 2021-01-02  2021-01   
2  Acre  Brazil      42046     802                 1.91 2021-01-03  2021-01   
3  Acre  Brazil      42117     806                 1.91 2021-01-04  2021-01   
4  Acre  Brazil      42170     808                 1.92 2021-01-05  2021-01   

   year  population  confirmed_1d  confirmed_moving_avg_7d   
0  2021      881935          <NA>                     <NA>  \
1  2021      881935           252                     <NA>   
2  2021      881935           105                     <NA>   
3  2021      881935            71                     <NA>   
4  2021      881935            53                     <NA>   

   confirmed_moving_avg_7d_rate_14d confirmed_trend  deaths_1d   
0                               NaN             NaN       <NA>  \
1                               NaN             NaN          2   
2                               NaN             NaN          4   
3                               NaN             NaN          4   
4                               NaN             NaN          2   

   deaths_moving_avg_7d  deaths_moving_avg_7d_rate_14d deaths_trend  
0                  <NA>                            NaN          NaN  
1                  <NA>                            NaN          NaN  
2                  <NA>                            NaN          NaN  
3                  <NA>                            NaN          NaN  
4                  <NA>                            NaN          NaN

Por fim, vamos reorganizar as colunas e conferir o resultado final.

In [20]:
cases = cases[['date', 'country', 'state', 'population', 'confirmed', 'confirmed_1d', 'confirmed_moving_avg_7d', 'confirmed_moving_avg_7d_rate_14d', 'confirmed_trend', 'deaths', 'deaths_1d', 'deaths_moving_avg_7d', 'deaths_moving_avg_7d_rate_14d', 'deaths_trend','case_fatality_ratio', 'month', 'year']]

In [21]:
cases.head(n=25)

date country state  population  confirmed  confirmed_1d   
0  2021-01-01  Brazil  Acre      881935      41689          <NA>  \
1  2021-01-02  Brazil  Acre      881935      41941           252   
2  2021-01-03  Brazil  Acre      881935      42046           105   
3  2021-01-04  Brazil  Acre      881935      42117            71   
4  2021-01-05  Brazil  Acre      881935      42170            53   
5  2021-01-06  Brazil  Acre      881935      42378           208   
6  2021-01-07  Brazil  Acre      881935      42478           100   
7  2021-01-08  Brazil  Acre      881935      42814           336   
8  2021-01-09  Brazil  Acre      881935      42908            94   
9  2021-01-10  Brazil  Acre      881935      43127           219   
10 2021-01-11  Brazil  Acre      881935      43346           219   
11 2021-01-12  Brazil  Acre      881935      43432            86   
12 2021-01-13  Brazil  Acre      881935      43785           353   
13 2021-01-14  Brazil  Acre      881935      44068           283   
14 2021-01-15  Brazil  Acre      881935      44156            88   
15 2021-01-16  Brazil  Acre      881935      44621           465   
16 2021-01-17  Brazil  Acre      881935      44767           146   
17 2021-01-18  Brazil  Acre      881935      44775             8   
18 2021-01-19  Brazil  Acre      881935      45208           433   
19 2021-01-20  Brazil  Acre      881935      45429           221   
20 2021-01-21  Brazil  Acre      881935      45729           300   
21 2021-01-22  Brazil  Acre      881935      45987           258   
22 2021-01-23  Brazil  Acre      881935      46239           252   
23 2021-01-24  Brazil  Acre      881935      46429           190   
24 2021-01-25  Brazil  Acre      881935      46539           110   

    confirmed_moving_avg_7d  confirmed_moving_avg_7d_rate_14d confirmed_trend   
0                      <NA>                               NaN             NaN  \
1                      <NA>                               NaN             NaN   
2                      <NA>                               NaN             NaN   
3                      <NA>                               NaN             NaN   
4                      <NA>                               NaN             NaN   
5                      <NA>                               NaN             NaN   
6                      <NA>                               NaN             NaN   
7                       161                               NaN             NaN   
8                       139                               NaN             NaN   
9                       155                               NaN             NaN   
10                      176                               NaN             NaN   
11                      181                               NaN             NaN   
12                      201                               NaN             NaN   
13                      228                               NaN             NaN   
14                      192                               NaN             NaN   
15                      245                               NaN             NaN   
16                      235                               NaN             NaN   
17                      205                               NaN             NaN   
18                      254                               NaN             NaN   
19                      235                               NaN             NaN   
20                      238                               NaN             NaN   
21                      262                          1.627329          upward   
22                      232                          1.669065          upward   
23                      238                          1.535484          upward   
24                      252                          1.431818          upward   

    deaths  deaths_1d  deaths_moving_avg_7d  deaths_moving_avg_7d_rate_14d   
0      796       <NA>                  <NA>            

#### **2.1.3. Carregamento** 

Com os dados manipulados, vamos persisti-lo em disco, fazer o seu download e carrega-lo no Google Data Studio.

In [22]:
cases.to_csv('./covid-cases.csv', sep=',', index=False)

### **2.2. Vacinação** 

Vamos processar os dados de **vacinação** da universidade de Oxford.

#### **2.2.1. Extração**

Os dados estão compilados em um único arquivo.

In [24]:
vaccines = pd.read_csv('https://covid.ourworldindata.org/data/owid-covid-data.csv', sep=',', parse_dates=[3])

In [25]:
vaccines.head()

iso_code continent     location       date  total_cases  new_cases   
0      AFG      Asia  Afghanistan 2020-01-03          NaN        0.0  \
1      AFG      Asia  Afghanistan 2020-01-04          NaN        0.0   
2      AFG      Asia  Afghanistan 2020-01-05          NaN        0.0   
3      AFG      Asia  Afghanistan 2020-01-06          NaN        0.0   
4      AFG      Asia  Afghanistan 2020-01-07          NaN        0.0   

   new_cases_smoothed  total_deaths  new_deaths  new_deaths_smoothed  ...   
0                 NaN           NaN         0.0                  NaN  ...  \
1                 NaN           NaN         0.0                  NaN  ...   
2                 NaN           NaN         0.0                  NaN  ...   
3                 NaN           NaN         0.0                  NaN  ...   
4                 NaN           NaN         0.0                  NaN  ...   

   male_smokers  handwashing_facilities  hospital_beds_per_thousand   
0           NaN                  37.746                         0.5  \
1           NaN                  37.746                         0.5   
2           NaN                  37.746                         0.5   
3           NaN                  37.746                         0.5   
4           NaN                  37.746                         0.5   

   life_expectancy  human_development_index  population   
0            64.83                    0.511  41128772.0  \
1            64.83                    0.511  41128772.0   
2            64.83                    0.511  41128772.0   
3            64.83                    0.511  41128772.0   
4            64.83                    0.511  41128772.0   

   excess_mortality_cumulative_absolute  excess_mortality_cumulative   
0                                   NaN                          NaN  \
1                                   NaN                          NaN   
2                                   NaN                          NaN   
3                                   NaN                          NaN   
4                                   NaN                          NaN   

   excess_mortality  excess_mortality_cumulative_per_million  
0               NaN                                      NaN  
1               NaN                                      NaN  
2               NaN                                      NaN  
3               NaN                                      NaN  
4               NaN                                      NaN  

[5 rows x 67 columns]

Vamos selecionar as colunas de interesse e as linhas referentes ao Brasil.

In [26]:
vaccines = vaccines.query('location == "Brazil"').reset_index(drop=True)
vaccines = vaccines[['location', 'population', 'total_vaccinations', 'people_vaccinated', 'people_fully_vaccinated', 'total_boosters', 'date']]

In [27]:
vaccines

location   population  total_vaccinations  people_vaccinated   
0      Brazil  215313504.0                 NaN                NaN  \
1      Brazil  215313504.0                 NaN                NaN   
2      Brazil  215313504.0                 NaN                NaN   
3      Brazil  215313504.0                 NaN                NaN   
4      Brazil  215313504.0                 NaN                NaN   
...       ...          ...                 ...                ...   
1268   Brazil  215313504.0                 NaN                NaN   
1269   Brazil  215313504.0                 NaN                NaN   
1270   Brazil  215313504.0                 NaN                NaN   
1271   Brazil  215313504.0                 NaN                NaN   
1272   Brazil  215313504.0                 NaN                NaN   

      people_fully_vaccinated  total_boosters       date  
0                         NaN             NaN 2020-01-03  
1                         NaN             NaN 2020-01-04  
2                         NaN             NaN 2020-01-05  
3                         NaN             NaN 2020-01-06  
4                         NaN             NaN 2020-01-07  
...                       ...             ...        ...  
1268                      NaN             NaN 2023-06-24  
1269                      NaN             NaN 2023-06-25  
1270                      NaN             NaN 2023-06-26  
1271                      NaN             NaN 2023-06-27  
1272                      NaN             NaN 2023-06-28  

[1273 rows x 7 columns]

#### **2.2.2. Wrangling**

Vamos manipular os dados para o dashboard. O foco é em garantir uma boa granularidade e qualidade da base de dados.

In [28]:
vaccines.head()

location   population  total_vaccinations  people_vaccinated   
0   Brazil  215313504.0                 NaN                NaN  \
1   Brazil  215313504.0                 NaN                NaN   
2   Brazil  215313504.0                 NaN                NaN   
3   Brazil  215313504.0                 NaN                NaN   
4   Brazil  215313504.0                 NaN                NaN   

   people_fully_vaccinated  total_boosters       date  
0                      NaN             NaN 2020-01-03  
1                      NaN             NaN 2020-01-04  
2                      NaN             NaN 2020-01-05  
3                      NaN             NaN 2020-01-06  
4                      NaN             NaN 2020-01-07

In [29]:
vaccines.shape

(1273, 7)

In [30]:
# É observável que os dados de vacinação no dataframe tem bastante valores nulos. Isso se deve pela vacinação ter começado tardiamente no Brasil.
vaccines.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1273 entries, 0 to 1272
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   location                 1273 non-null   object        
 1   population               1273 non-null   float64       
 2   total_vaccinations       695 non-null    float64       
 3   people_vaccinated        691 non-null    float64       
 4   people_fully_vaccinated  675 non-null    float64       
 5   total_boosters           455 non-null    float64       
 6   date                     1273 non-null   datetime64[ns]
dtypes: datetime64[ns](1), float64(5), object(1)
memory usage: 69.7+ KB


Vamos começar tratando os dados faltantes, a estratégia será a de preencher os buracos com o valor anterior válido mais próximo.

In [31]:
vaccines = vaccines.fillna(method='ffill')

Vamos também filtrar a base de dados de acordo com a coluna `date` para garantir que ambas as bases de dados tratam do mesmo período de tempo.

In [32]:
vaccines = vaccines[(vaccines['date'] >= '2021-01-01') & (vaccines['date'] <= '2022-12-31')].reset_index(drop=True)

Agora, vamos alterar o nome das colunas.

In [33]:
vaccines = vaccines.rename(
  columns={
    'location': 'country',
    'total_vaccinations': 'total',
    'people_vaccinated': 'one_shot',
    'people_fully_vaccinated': 'two_shots',
    'total_boosters': 'three_shots',
  }
)

Vamos então computar novas colunas para enriquecer a base de dados.

 - Chaves temporais:

In [34]:
vaccines['month'] = vaccines['date'].apply(lambda date: date.strftime('%Y-%m'))
vaccines['year']  = vaccines['date'].apply(lambda date: date.strftime('%Y'))

 - Dados relativos:

In [35]:
vaccines['one_shot_perc'] = round(vaccines['one_shot'] / vaccines['population'], 4)
vaccines['two_shots_perc'] = round(vaccines['two_shots'] / vaccines['population'], 4)
vaccines['three_shots_perc'] = round(vaccines['three_shots'] / vaccines['population'], 4)

Garantir o tipo do dado é fundamental para consistência da base de dados. Vamos fazer o *type casting* das colunas.

In [36]:
vaccines['population'] = vaccines['population'].astype('Int64')
vaccines['total'] = vaccines['total'].astype('Int64')
vaccines['one_shot'] = vaccines['one_shot'].astype('Int64')
vaccines['two_shots'] = vaccines['two_shots'].astype('Int64')
vaccines['three_shots'] = vaccines['three_shots'].astype('Int64')

Por fim, vamos reorganizar as colunas e conferir o resultado final.

In [37]:
vaccines = vaccines[['date', 'country', 'population', 'total', 'one_shot', 'one_shot_perc', 'two_shots', 'two_shots_perc', 'three_shots', 'three_shots_perc', 'month', 'year']]

In [38]:
vaccines

date country  population      total   one_shot  one_shot_perc   
0   2021-01-01  Brazil   215313504       <NA>       <NA>            NaN  \
1   2021-01-02  Brazil   215313504       <NA>       <NA>            NaN   
2   2021-01-03  Brazil   215313504       <NA>       <NA>            NaN   
3   2021-01-04  Brazil   215313504       <NA>       <NA>            NaN   
4   2021-01-05  Brazil   215313504       <NA>       <NA>            NaN   
..         ...     ...         ...        ...        ...            ...   
725 2022-12-27  Brazil   215313504  480310839  188549744         0.8757   
726 2022-12-28  Brazil   215313504  480310839  188549744         0.8757   
727 2022-12-29  Brazil   215313504  480331769  188552661         0.8757   
728 2022-12-30  Brazil   215313504  480332769  188553047         0.8757   
729 2022-12-31  Brazil   215313504  480333910  188553932         0.8757   

     two_shots  two_shots_perc  three_shots  three_shots_perc    month  year  
0         <NA>             NaN         <NA>               NaN  2021-01  2021  
1         <NA>             NaN         <NA>               NaN  2021-01  2021  
2         <NA>             NaN         <NA>               NaN  2021-01  2021  
3         <NA>             NaN         <NA>               NaN  2021-01  2021  
4         <NA>             NaN         <NA>               NaN  2021-01  2021  
..         ...             ...          ...               ...      ...   ...  
725  174881292          0.8122    122616211            0.5695  2022-12  2022  
726  174881292          0.8122    122616211            0.5695  2022-12  2022  
727  174886102          0.8122    122629436            0.5695  2022-12  2022  
728  174886846          0.8122    122629436            0.5695  2022-12  2022  
729  174887915          0.8122    122629436            0.5695  2022-12  2022  

[730 rows x 12 columns]

#### **2.2.3. Carregamento** 

Com os dados manipulados, vamos persisti-lo em disco, fazer o seu download e carrega-lo no Google Data Studio.

In [39]:
vaccines.to_csv('./covid-vaccines.csv', sep=',', index=False)

## 3\. Exploração Interativa de Dados

### **3.1. KPIs** 

O dashboard de dados contem os seguintes indicadores chaves de desempenho (*key performance indicator* ou KPI) consolidados:

1. Casos e mortes nas 24 horas;
1. Média móvel (7 dias) de casos e mortes;
1. Tendência de casos e mortes;
1. Proporção de vacinados com 1ª, 2ª e 3ª doses.


### **3.2. EDA** 

O dashboard de dados contem os seguintes gráficos para a análise exploratória de dados (*exploratory data analysis*
ou EDA) interativa:

1. Distribuição do números de casos e mortes ao longo do tempo;
1. Distribuição da média móvel (7 dias) do números de casos e mortes ao longo do tempo;
1. Distribuição de risco de mortalidade média;
1. Distribuição geográfica dos casos por estado por dia.

## 4\. Storytelling

A partir da análise do dashboard, é possivel retirar algumas conclusões: 

- O estado com a maior média de fatalidade (quantidade de mortes por casos) é o RJ com 4,45%, seguindo de São Paulo.
- Apesar de haver uma queda no 2º semestre de 2021, houve um aumento gigantesco no número de casos confirmados em Janeiro e fevereiro de 2022. Isso pode ter váriados motivos, como a subnotificação e o aumento de testagem. Mas, a Organização Pan-Americana de Saúde ([link](https://www.paho.org/pt/noticias/16-2-2022-relaxamento-medidas-saude-publica-contribuiu-para-aumento-mortes-por-covid-19)) informa que esse aumento é devido as flexibilizações da população enquanto as politicas de lockdown. Seguindo o dashboard, isso faz sentido, pois a partir de Janeiro de 2022 já havia uma taxa de vacinação acima de 75%, o que levou as pessoas a tomarem maior liberdade para sairem de casa.
- A taxa de vacinação foi relativamente rápida, considerando que em menos de 1 ano, mais de 75% da população já tinha tomado a 1º dose e mais de 65% a 2º dose.